### PR TITLE
Build with c++14

### LINKED ISSUE #847: c++14 upgrade
The project extensively uses C++11 features and compiles with C++11 support. But it's 2018 now and I frankly don't see why not upgrade at least to C++14 (or even C++17). And it's quite easy to do: change a couple of cmake files and perhaps a few compiler warnings. The real question: are there any serious enough reasons to stick to C++11? 

We can do the upgrade and testing ourselves, if the community agrees and provides the list of desired platforms/compilers.

### PR DESCRIPTION
**Pull request by jmjatlanta**
Fixes #847 

The same change was made in the fc subproject at https://github.com/bitshares/bitshares-fc/pull/114 . Hence, fc should be bumped when it is merged in.

ToDo:

 - [ ] Update documentation to require CMake 3.1
 - [ ] Update documentation to require a compiler that supports c++14

**Note** c++14 has been marked "required" in the CMakeLists.txt. This requirement could be loosened, but would later break clients on very old compilers as soon as we use a c++14 feature. I do not know which is worse, finding those users now or later.

This has been tested on Ubuntu 16.04 / g++ 5.4.0 / cmake 3.5.1 (which are the default versions for Ubuntu 16.04).

This has been tested on macOS Mojave 10.14.3 with clang 1000.11.45.5 (LLVM 10.0.0) and cmake 3.14.0

This has been tested on Windows 10 with MSVC 19.16.27027.1 ( Visual Studio 15 2017 ) and cmake 3.14

### CODE REVIEW

**Comment by pmconrad** [comment]
We have not yet decided if we want to make c++14 mandatory.

**Comment by jmjatlanta** [comment]
It has been decided to begin using c++14 in release 3.2.0 of BitShares Core. This PR modifies CMakeLists.txt to tell the compiler to use c++14.

**Comment by pmconrad** [comment]
Please rebase instead of merging, the diff contains too many unrelated changes.

**Comment by jmjatlanta** [comment]
Sometimes I am so thick-headed. Sorry. Agreed. This PR is now almost impossible to review! I will fix it.

**Comment by pmconrad** [comment]
Please add changes in fc and bump, those in https://github.com/bitshares/bitshares-fc/pull/114 were not sufficient

**Comment by pmconrad** [comment]
Also, it would be good to have a small C++-14 only change in both fc and core to verify that the settings are complete and have the intended effect.

File: CMakeLists.txt [hunk 1]
```diff
@@ -69,10 +73,6 @@ IF(NOT "${Boost_VERSION}" MATCHES "1.53(.*)")
    SET(Boost_LIBRARIES ${BOOST_LIBRARIES_TEMP} ${Boost_LIBRARIES})
 ENDIF()
 
-if( NOT CPP_STANDARD )
-  set( CPP_STANDARD "-std=c++11" )
```
Comment (abitmore) [inline CMakeLists.txt]: This PR didn't work for me (Ubuntu 16.04, Gcc 5.4.0, FC revision 8176f39 (latest master)). Changing this line to `c++14` got me further (failed on `wallet.cpp`).

Comment (abitmore) [reply CMakeLists.txt]: Update: after changed `c++11` to `c++14` here, and changed `identity_member` to `identity_member_with_optionals` in wallet.cpp, my build worked. Should I create another PR?

Comment (abitmore) [reply CMakeLists.txt]: Update: I made a mistake, missed the `CMAKE_CXX_STANDARD 14 ` change, I think that's why I failed to build. Now rebuilding.

Comment (jmjatlanta) [reply CMakeLists.txt]: I fired up Ubuntu 16.04.04, this branch, and was able to compile fine.

Comment (abitmore) [reply CMakeLists.txt]: I guess you didn't bump FC.

@pmconrad: how about merging this PR as is, then bump FC in https://github.com/bitshares/bitshares-core/pull/1695?

**Pull request merged**